### PR TITLE
Build native assets by default

### PR DESCRIPTION
**Pull request by dougbu**
- #22556
- make `-BuildNative` primarily useful when you want _only_ native assets
- remove `-ForceCoreMsbuild` option

nit: extra `Remove-Item`s caused `MSBuild` function to do redundant work

### CODE REVIEW

File: build.ps1 [hunk 1]
```diff
@@ -195,11 +205,9 @@ if ($Projects) {
 }
 # When adding new sub-group build flags, add them to this check.
 elseif (-not ($All -or $BuildNative -or $BuildManaged -or $BuildNodeJS -or $BuildInstallers -or $BuildJava)) {
-    Write-Warning "No default group of projects was specified, so building the 'managed' and its dependent subsets of projects. Run ``build.cmd -help`` for more details."
-
-    # This goal of this is to pick a sensible default for `build.cmd` with zero arguments.
-    # Now that we support subfolder invokations of build.cmd, we will be pushing to have build.cmd build everything (-all) by default
```
Comment (dougbu) [inline build.ps1:201]: No need to track TODO items in the code. dotnet/aspnetcore#6304 tracks this.

File: eng/scripts/GenerateReferenceAssemblies.ps1 [hunk 2]
```diff
@@ -10,9 +10,6 @@ try {
   & "$repoRoot\build.ps1"  -ci:$ci -nobl -noBuildRepoTasks -noRestore -buildNative -configuration Debug
 
   Remove-Item variable:global:_BuildTool -ErrorAction Ignore
-  Remove-Item variable:global:_DotNetInstallDir -ErrorAction Ignore
-  Remove-Item variable:global:_ToolsetBuildProj -ErrorAction Ignore
-  Remove-Item variable:global:_MSBuildExe -ErrorAction Ignore
```
Comment (dougbu) [inline eng/scripts/GenerateReferenceAssemblies.ps1:15]: Removing redundant removals, as mentioned in the description

**Comment by dougbu** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=674531 demonstrated the success of this change in addressing #22556. The Windows build worked fine though the `dotnet test`s failed miserably (perhaps they need `--no-build`❔).

**Comment by dougbu** [comment]
After my update:
- [20200607.1](https://dev.azure.com/dnceng/internal/_build/results?buildId=675255&view=results) run of aspnetcore-blazor-daily-tests failed about the same way; build was fine `dotnet test` not so much
- [20200607.1](https://dev.azure.com/dnceng/public/_build/results?buildId=675252&view=results) succeeded and demonstrated more of our build scripts are working fine these days.

File: src/Servers/build.cmd [hunk 3]
```diff
@@ -1,3 +1,3 @@
 @ECHO OFF
 SET RepoRoot=%~dp0..\..
-%RepoRoot%\build.cmd -projects %~dp0**\*.*proj %*
+%RepoRoot%\build.cmd -projects %~dp0**\*.csproj %*
```
Comment (wtgodbe) [inline src/Servers/build.cmd:3]: What's the purpose of this change?

**Review by wtgodbe — APPROVED** [review]

**Comment by jkotalik** [comment]
> make -BuildNative primarily useful when you want only native assets

Can you clarify this statement a bit more? So now calling `build.cmd -BuildNative` will only build native bits?

I think this breaks the build.cmd script in the IIS folder: https://github.com/dotnet/aspnetcore/blob/master/src/Servers/IIS/build.cmd. Could you adjust that accordingly?

File: src/Servers/build.cmd (hunk 3 above)
Comment (dougbu) [reply src/Servers/build.cmd:3]: The `-projects` list isn't split between native and non-native projects. That means `**\*.*proj` would be passed to `dotnet msbuild` and blow up spectacularly because only desktop `msbuild` knows how to build the native projects. Further, because this does **not** pass `-noBuildNative` the native projects are always built.

**Comment by dougbu** [comment]
@jkotalik good catch. That command previously worked because desktop `msbuild` would handle any `*.vcxproj` dependencies of the managed projects listed. The `$(BuildNative)` setting made sure those dependencies weren't ignored. Now, the script will attempt to build the managed projects using desktop `msbuild` and fail unless the system has VS 16.6.0 (actually preview 2 IIRC) or later.

Fix is simply to remove `-buildNative` from the command line. Will do before I merge.

**Comment by dougbu** [comment]
> So now calling `build.cmd -BuildNative` will only build native bits?

That command has always done that. The differences are
1. `.\build.cmd -projects {some projects}` will build the projects using `dotnet msbuild` while `.\build.cmd -buildNative {some projects}` will build them using desktop `msbuild`
2. `.\build.cmd` will build both managed and native projects because native projects are built _except_ if you specify `-noBuildNative`

**Comment by jkotalik** [comment]
> .\build.cmd will build both managed and native projects because native projects are built except if you specify -noBuildNative

Does that mean you need to have the VS toolsets necessary to build native for build.cmd to succeed? What happens if I call build.cmd without having it installed?

We made a very conscious decision to make it so build.cmd wouldn't build native dlls, as it isn't easy to have everything installed and slows down the build process.

**Comment by dougbu** [comment]
@jkotalik the answer depends on the specific build.cmd script you run. The problem without the fixes we've made over the week or so (starting with 14d6b6e28661) is builds using desktop `msbuild` failed miserably with VS 2019 16.5 or earlier failed because that `msbuild` didn't understand `net5.0`.

We're not in the perfect place with this. But, it's better than those failures or (worse) having to run `.\build.cmd` multiple times to pick up the native assets needed in the shared framework i.e. to get the CI builds working.

In any case, those who don't care about building the shared framework properly or aren't testing with IIS can execute `./build.cmd -noBuildNative`. The native build takes about 3 minutes for the rest of us.

**Comment by jkotalik** [comment]
> In any case, those who don't care about building the shared framework properly or aren't testing with IIS can execute ./build.cmd -noBuildNative. The native build takes about 3 minutes for the rest of us.

I still think this should be opt in to building native rather than opt out. Imagine a new developer (ex external customer) to the aspnetcore repo tries to run build.cmd from the root or a subfolder. It fails due to native components not being installed in VS. They need to now install dependencies separately and then run build.cmd again. Making build.cmd not build native helped with onboarding making changes in the repo. Also, developers making changes to the aspnetcore repo rarely need to build native dependencies.

>  The native build takes about 3 minutes for the rest of us.

From my experience, it takes significantly longer than 3 minutes on CI to build native. I was seeing ~15 minutes when I was testing ways to speed up build times for our CI.

**Comment by dougbu** [comment]
> From my experience, it takes significantly longer than 3 minutes on CI to build native.

Native build takes ~6 minutes on the CI agents and those are significantly underpowered compared to almost any dev machine. See https://dev.azure.com/dnceng/internal/_build/results?buildId=676527&view=logs&j=1b89928a-2219-5ef9-602f-f95beb3da4dc&t=bcce746a-741e-59b7-adec-a4a82d8c5e52&l=316 for example.

In any case, these builds are required on the CI, are done once per build job, and isn't relevant to your point about dev happiness.

> I still think this should be opt in to building native rather than opt out

I'm not so sure. We also have issue #6304 about having the root build.cmd script building everything.

If we do something here, would be separate from `-buildNative` because that has the side effect of changing how `-projects` is interpreted. Maybe `-includeNativePhase`❔ One concern would be the need to change a bunch of the YAML and build.cmd|ps1 files to get the CI working again ☹️

**Comment by jkotalik** [comment]
> Native build takes ~6 minutes on the CI agents and those are significantly underpowered compared to almost any dev machine. See https://dev.azure.com/dnceng/internal/_build/results?buildId=676527&view=logs&j=1b89928a-2219-5ef9-602f-f95beb3da4dc&t=bcce746a-741e-59b7-adec-a4a82d8c5e52&l=316 for example.

Wow, that's a lot faster than it was before. That's great!

> I'm not so sure. We also have issue #6304 about having the root build.cmd script building everything.

If that's the direction for the root build.cmd, I think that's fine. I think I'm most concerned about the subdirectory build.cmd calls and how they are affected. For example, what happens when calling src/Servers/Kestrel/build.cmd when someone doesn't have the native dependencies installed? Does the build fail? That's my main concern.

**Comment by dougbu** [comment]
> what happens when calling src/Servers/Kestrel/build.cmd when someone doesn't have the native dependencies installed?

build.ps1 isn't smart enough to notice the user lacks the needed `msbuild` components to build native projects. So, it'll run desktop `msbuild` with `/p:BuildNative=true` that global property means the [override here](https://github.com/dotnet/aspnetcore/blob/master/eng/Common.props#L14) doesn't work and the build will fail.

This brings to mind another solution option if we decide the experience is significantly degraded: We could figure out a way to skip the native phase when the needed `msbuild` components are missing, a bit like what we do when NodeJS isn't found. The other option is to fail up front, a lot like we do when Java isn't found. (Notice the inconsistency there ☹️)

@jkotalik should we move this to a new issue on the DoI board❔ I suspect we want to change something somewhere and am not surprised given how quickly we've pivoted to using `dotnet msbuild` by default then cleaning up.

**Comment by jkotalik** [comment]
>  We could figure out a way to skip the native phase when the needed msbuild components are missing, a bit like what we do when NodeJS isn't found. 

I'd be okay with that option. We could briefly bring this up at DoI as well.
